### PR TITLE
feat(AsyncPipe): allow onError argument

### LINKED ISSUE #7539: feat(AsyncPipe): allow for error and/or complete callbacks as arguments to |async pipe
```html
<h3>Repo list</h3>
<ul>
	<li *ngFor="#repo of repos | async:onError">
    <a [routerLink]=" ['RepoDetail', {org: repo.owner.login, name: repo.name}] ">
      {{ repo.name }}
    </a>
  </li>
</ul>
```

* **I'm submitting a ... **
[ ] bug report
[x] feature request
[ ] support request => Please do not submit support request here, see note at the top of this template.

* **Do you want to request a *feature* or report a *bug*?**
feature

* **What is the current behavior?**
when there's an error the observable error throws

* **What is the expected behavior?**
some users wants a graceful way of handling the error

* **What is the motivation / use case for changing the behavior?**
even though you would manually subscribe to the observable for more control when prototyping or for demos etc it's best to use the abstraction

### PR DESCRIPTION
**Pull request by PatrickJS**
- **Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)
- **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
  feature
- **What is the current behavior?** (You can also link to an open issue here)
  an error is thrown within the template 
- **What is the new behavior (if this is a feature change)?**
  allow passing a onError callback
- **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
  nope
- **Other information**:
  closes #7539

### CODE REVIEW

**Comment by robwormald** [comment]
what's the usage look like? something like
```
<div *ngFor="#foo of foos | async:onError">
```
?

**Comment by gdi2290** [comment]
@robwormald ya

**Comment by mhevery** [comment]
Hey @robwormald  & @gdi2290,

It seems a bit strange to me that one would put error handling in the UI.

1. What is the current behavior in case of an error? 
2. What is the motivation for this? What would one do with this error other the report it?
3. Could we think of a way to do this in a way which is more component scoped? As it a component could say 'I would like to get all async pipe errors delivered here'

PS: Just trying to get a better feel for the motivation behind thes PR.

**Comment by Martin-Wegner** [comment]
Any plans to fix this PR?

File: modules/angular2/src/common/pipes/async_pipe.ts [hunk 1]
```diff
@@ -1,4 +1,5 @@
 import {isBlank, isPresent, isPromise, CONST} from 'angular2/src/facade/lang';
+import {ListWrapper} from 'angular2/src/facade/collection';
```
Comment (gdi2290) [inline modules/angular2/src/common/pipes/async_pipe.ts]: we don't need this anymore

**Comment by stevenjob** [comment]
Why was this closed?
The asynchronous pipe seems useless without a way to handle errors

**Comment by DzmitryShylovich** [comment]
@stevenjob Observable has `.catch()` method

**Comment by stevenjob** [comment]
doesn't catch require you to return an observable?
In reality you would want to show an error message or something?

**Comment by DzmitryShylovich** [comment]
You can use if/else style syntax:
```
<ng-template #error>Error!</ng-template>
<div *ngIf="userObservable | async as user; else error">
  {{ user.name }}
</div>

// inside component
userObservable = this.http.get('user').catch(error => Observable.of(false));
```

**Comment by stevenjob** [comment]
Returning an observable of false from a stream that is supposed to return user would break typescript and seems like a smell to me.
Also i imagine the `else` here would be used to check for loading when the stream has not returned a value yet and would be falsy?

I have seen people use `-1` for error value but its not a nice solution.

```
// inside component would get `Type Observable<bool> is not assignable to Observable<User> ...`
userObservable: Observable<User> = this.http.get('user').catch(error => Observable.of(false));
```


Since the async pipe is doing the same job as using subscribe it should handle all the same things as subscribe e.g. onNext, onError, onCompleted


To handle all cases:
```

<ng-template #loading>Loading!</ng-template>
<div *ngIf="userObservable | async:onError:onCompleted as user; else loading">
  {{ user.name }}
</div>

// inside component
userObservable = this.http.get('user');


public onError(error) {
  // do something with error
}

public onCompleted() {
  // do something
}
```

**Comment by DzmitryShylovich** [comment]
> Returning an observable of false from a stream that is supposed to return user would break typescript and seems like a smell to me.

ok, you can return null/undefined

> Since the async pipe is doing the same job as using subscribe it should handle all the same things as subscribe e.g. onNext, onError, onCompleted

if you need all 3 callbacks just subscribe inside a component

**Comment by lordscales91** [comment]
So the conclusion is: never use async pipes if you want to handle errors properly, specially if you want to display an error message sent from the server.

**Comment by rgant** [comment]
A problem with AsyncPipe throwing errors is that it makes them uncatchable in jasmine tests as far as I can tell. If you are testing the error handling then you want the errors to make it to the template, but you don't want those errors to reach jasmine and cause the test to fail with Uncaught error. I had to work around this issue by patch AsyncPipe to not throw errors in my tests. 

https://stackoverflow.com/q/52433159/41908

**Pull request closed without merging**